pkg/envoy/server.go

```diff
@@ -506,22 +506,46 @@ func getNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
 
 // UpdateNetworkPolicy adds or updates a network policy in the set of published
 // to L7 proxies.
-func UpdateNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+// When the proxy acknowledges the network policy update, it will result in
+// a subsequent call to the endpoint's OnProxyPolicyAcknowledge() function.
+func UpdateNetworkPolicy(ep NetworkPolicyEndpoint, policy *policy.L4Policy,
 	labelsMap identity.IdentityCache, deniedIngressIdentities, deniedEgressIdentities map[identity.NumericIdentity]bool) error {
-	networkPolicy := getNetworkPolicy(id, policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities)
-	err := networkPolicy.Validate()
-	if err != nil {
-		return fmt.Errorf("error validating generated NetworkPolicy: %s", err)
+
+	// First, validate all policies
+	ips := []string{
+		ep.GetIPv6Address(),
+		ep.GetIPv4Address(),
+	}
+	policies := []*cilium.NetworkPolicy{}
+	for i := range ips {
+		if ips[i] == "" {
+			continue
+		}
+		networkPolicy := getNetworkPolicy(ips[i], ep.GetIdentity(), policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities)
+		err := networkPolicy.Validate()
+		if err != nil {
+			return fmt.Errorf("error validating generated NetworkPolicy for %s: %s", ips[i], err)
+		}
+		policies = append(policies, networkPolicy)
 	}
 
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, name, networkPolicy, false)
+	// When successful, push them into the cache.
+	unackedNPDSUpdates.Insert(ep, policy.Revision, NetworkPolicyCache.GetVersion())
```

Comment:
Call `wg.AddCompletion()` to get a `Completion` and pass that to the `Insert` call to store it in the binding.